platform/windows/display_server_windows.cpp

```diff
@@ -95,6 +95,98 @@ static void track_mouse_leave_event(HWND hWnd) {
 	TrackMouseEvent(&tme);
 }
 
+class WindowsNativeIcons {
+	struct IconEntry {
+		int width;
+		HICON hIcon;
+	};
+	Vector<IconEntry> icons;
+
+public:
+	WindowsNativeIcons() = default;
+	~WindowsNativeIcons();
+	WindowsNativeIcons(const WindowsNativeIcons &) = delete;
+	WindowsNativeIcons &operator=(const WindowsNativeIcons &) = delete;
+
+	bool load_native_icons(const Ref<FileAccess> &ico_file);
+	bool load_from_image(const Ref<Image> &img);
+	void clear_icons();
+
+	bool is_valid() const { return !icons.is_empty(); }
+	HICON icon_for_size(int width) const;
+
+	static int icon_size_small(int dpi = 0);
+	static int icon_size_big(int dpi = 0);
+};
+
+WindowsNativeIcons::~WindowsNativeIcons() {
+	clear_icons();
+}
+
+void WindowsNativeIcons::clear_icons() {
+	for (IconEntry &icon : icons) {
+		DestroyIcon(icon.hIcon);
+	}
+	icons.clear();
+}
+
+HICON WindowsNativeIcons::icon_for_size(int width) const {
+	HICON hIcon = nullptr;
+	for (const IconEntry &icon : icons) {
+		hIcon = icon.hIcon;
+		if (icon.width >= width) {
+			break;
+		}
+	}
+	return hIcon;
+}
+
+int WindowsNativeIcons::icon_size_small(int dpi) {
+	if (dpi == 0) {
+		return GetSystemMetrics(SM_CXSMICON);
+	}
+
+	using GetSystemMetricsForDpiPtr = int (*)(int nIndex, UINT dpi);
+	static GetSystemMetricsForDpiPtr GetSystemMetricsForDpi = (GetSystemMetricsForDpiPtr)GetProcAddress(LoadLibraryW(L"user32.dll"), "GetSystemMetricsForDpi");
+	if (GetSystemMetricsForDpi) {
+		return GetSystemMetricsForDpi(SM_CXSMICON, dpi);
+	}
+	return 16 * dpi / 96;
+}
+
+int WindowsNativeIcons::icon_size_big(int dpi) {
+	static bool is_win10 = []() {
```

Comment:
Usually, we only allow lambdas if it's necessary for performance or readability, see https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#lambdas. Neither is the case here.